Enable NETC support for i.MX95 EKV M7

yangbolu1991 opened the pull request:
This PR is to enable NETC support for i.MX95 EKV M7.
There are dependencies of this PR.
- hal_nxp: https://github.com/zephyrproject-rtos/hal_nxp/pull/462
- scmi: https://github.com/zephyrproject-rtos/zephyr/pull/77437

Current PR already cherry-pick scmi changes and updated west.yml to address the dependency for CI.
And will drop them once they are merged.

Thanks.

yangbolu1991 committed:
dts: arm: nxp_imx95_m7: add power domain node and definitions

Added power domain node and i.MX95 power domain header file.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
dts: arm: nxp_imx95_m7: add NETC node and IRQSTEER node

Added NETC device tree node, and IRQSTEER node.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
soc: nxp: imx95: support M7 soc init for NETC

Added support for soc init for NETC.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: imx95_evk: add board init for NETC

Added board init for NETC.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: imx95_evk: enable NETC PSI0 for M7

Enabled NETC PSI0 for M7.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: eth_nxp_imx_netc: get irq from dts node

Added support to get irq from dts node, because some platforms may
use multi-level interrupts, and the system INTID may not be connected
directly to interrupt controller.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v2. Changes,

- Rebased on latest scmi patches. And addressed west.yml conflict.

Thanks.

yangbolu1991 committed:
soc: nxp: imx95: enable multi-level interrupts for m7

Enabled multi-level interrupts for m7 since IRQSTEER is used.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

mmahadevan108 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -6,6 +6,7 @@
 
 #include <arm/armv7-m.dtsi>
 #include <zephyr/dt-bindings/clock/imx95_clock.h>
+#include <zephyr/dt-bindings/power/imx95_power.h>
```
I don't see any `macro`  from this header used. Is this needed?

yangbolu1991 replied:
Let me remove it. It's true IP node hadn't used it now.
It's used in soc.c for power init now.

mmahadevan108 commented:
Please rebase and update west.yml

mmahadevan108 reviewed (changes requested):
Needs to be sequenced behind https://github.com/zephyrproject-rtos/zephyr/pull/83827

@yangbolu1991  I merged the HAL side so please update your PR to point to the new SHA. When the gating PR merges I'll clear my CR

mmahadevan108 reviewed (dismissed):
I have rebased this PR.

mmahadevan108 commented:
Please help address the CI failures.

yangbolu1991 commented:
Sory, I have an update on the PR.

Updated to v3. Changes,

- Converted to get irq for NETC from dts node instead of IRQ MACRO defined in device header. This required irqsteer driver support. So it depends on https://github.com/zephyrproject-rtos/zephyr/pull/84017
- Updated hal_nxp for some fix-up. Depend on https://github.com/zephyrproject-rtos/hal_nxp/pull/499
- Removed imx95_power.h including in dtsi as it's not used now.

Thank you very much.

yangbolu1991 committed:
tests: arm: exclude i.mx95 m7 for some interrupt tests

i.MX95 M7 is using multi-level interrupts with custom
interrupt handling. So, arm_custom_interrupt and
arm_irq_vector_table tests should exclude it.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: imx95_evk: update doc and yml for NETC support on M7

Updated imx95_evk doc and yml for NETC support on M7.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v4. Changes include,

- Rebased.
- Added one more patch to exclude tests not suit for multilevel interrupts.

Dependency (ported here for CI):
- https://github.com/zephyrproject-rtos/zephyr/pull/84017
- https://github.com/zephyrproject-rtos/hal_nxp/pull/499

Thank you very much.

yangbolu1991 commented:
Updated to v5. Changes include,
- Rebased as some changes already merged, and the dependency PR had update.

Dependency:
- https://github.com/zephyrproject-rtos/zephyr/pull/84017
- https://github.com/zephyrproject-rtos/hal_nxp/pull/499

Thank you very much.

dleach02 commented:
@yangbolu1991 please rebase your PR

mmahadevan108 commented:
@yangbolu1991, please rebase and update `west.yml`

yangbolu1991 committed:
west.yml: update hal_nxp to support i.MX95 NETC

Updated hal_nxp to support i.MX95 NETC.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v6. Changes,
- Rebased.

But It looks like there is twister issues on code base.

Thanks.

yangbolu1991 commented:
Rebased to try CI again.
Thanks.

yangbolu1991 commented:
Rebased, thanks.

mmahadevan108 commented:
@JiafeiPan , can you help review this PR.

mmahadevan108 commented on boards/nxp/imx95_evk/board.c:
```diff
@@ -0,0 +1,36 @@
+#include <soc.h>
+
+static int board_init(void)
+{
```
Please look at the changes made as part of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/85605 and update this implementation accordingly.

yangbolu1991 replied:
Actually the soc/board early init hooks are not able to be used for the SoC based on SCMI APIs.
I had commented in file.

```
/*
 * Because platform is using ARM SCMI, drivers like scmi, mbox etc. are
 * initialized during PRE_KERNEL_1. Common init hooks is not able to use.
 * SoC early init and board early init could be run during PRE_KERNEL_2 instead.
 */
```

JiafeiPan commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_m7.dts:
```diff
@@ -23,6 +23,27 @@
+	pinctrl-names = "default";
+	phy-handle = <&phy0>;
+	phy-connection-type = "rgmii";
+	status = "okay";
```
I suggest to disable all related nodes by default in board dts, and enable them in dts overlay, then can use "`-DEXTRA_DTC_OVERLAY_FILE=enet1.overlay`" if needed.

yangbolu1991 replied:
Thanks Jiafei. Will do that.

yangbolu1991 commented:
Updated to v7. Changes include,
- Created overlay file for enetc psi0 instead, and doc it.

Thanks.

JiafeiPan commented on boards/nxp/imx95_evk/enetc_psi0.overlay:
```diff
@@ -0,0 +1,17 @@
+/*
```
We may have the other overlay for other features, I suggest to put all of them in "dts" sub-directory just like what the other platforms are using.

JiafeiPan commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_m7.dts:
```diff
@@ -23,6 +23,24 @@
 
+&emdio {
+	pinctrl-0 = <&emdio_default>;
+	pinctrl-names = "default";
```
I know these nodes are disabled in soc dts, but I prefer to put explict "disabled" status for these nodes to make it more readable for the user to know it is disabled currently.

yangbolu1991 replied:
okay, updated.
Thanks.

yangbolu1991 commented:
Updated to v8. Changes include,
- moved overlay file to dts/ directory.
- put explicit disabled status in dts.

Thanks.
